File: test_candidate_payments.js
/**
 * Integration test for candidate payment system
 *
 * Tests:
 *   1. candidate-pay requires candidato role (admin → 403)
 *   2. candidate-pay requires group_exam_id (400)
 *   3. candidate-pay requires token (400)
 *   4. candidate-pay with invalid group_exam_id (404)
 *   5. candidate-retake requires candidato role (403)
 *   6. candidate-retake with invalid group_exam_id (404)
 *   7. mis-examenes includes payment fields
 *   8. check-access includes payment fields
 *   9. my-payments includes payment_type field
 *   10. editor cannot use candidate endpoints (403)
 *
 * Run: node test_candidate_payments.js [dev|prod]
 */
const https = require('https');

const ENV = process.argv[2] || 'dev';
const API = ENV === 'prod'
  ? 'https://evaluaasi-motorv2-api.purpleocean-384694c4.southcentralus.azurecontainerapps.io/api'
  : 'https://evaluaasi-motorv2-api-dev.purpleocean-384694c4.southcentralus.azurecontainerapps.io/api';

function request(method, path, token, body) {
  return new Promise((resolve, reject) => {
    const url = new URL(API + path);
    const opts = {
      hostname: url.hostname, path: url.pathname + url.search, method,
      headers: { 'Content-Type': 'application/json' },
    };
    if (token) opts.headers['Authorization'] = `Bearer ${token}`;
    const req = https.request(opts, (res) => {
      let data = '';
      res.on('data', (c) => (data += c));
      res.on('end', () => {
        try { resolve({ status: res.statusCode, data: JSON.parse(data) }); }
        catch { resolve({ status: res.statusCode, data }); }
      });
    });
    req.on('error', reject);
    if (body) req.write(JSON.stringify(body));
    req.end();
  });
}

async function login(username, password, retries = 3) {
  for (let i = 0; i < retries; i++) {
    const res = await request('POST', '/auth/login', null, { username, password });
    if (res.data && res.data.access_token) return res.data.access_token;
    if (i < retries - 1) await new Promise(r => setTimeout(r, 5000));
  }
  throw new Error(`Login failed for ${username}`);
}

let passed = 0, failed = 0;
function assert(condition, msg) {
  if (condition) { passed++; console.log(`  PASS: ${msg}`); }
  else { failed++; console.log(`  FAIL: ${msg}`); }
}

async function main() {
  console.log(`=== Candidate Payments Test — ${ENV.toUpperCase()} ===`);
  console.log(`API: ${API}\n`);

  // ── Login ──
  console.log('1. Logging in...');
  const adminToken = await login('admin', 'admin123');
  console.log('   Admin login: OK');

  let alumnoToken;
  try {
    alumnoToken = await login('U343Z793S8', 'TestPay2024!');
    console.log('   Candidato login: OK');
  } catch (e) {
    console.log('   Candidato login FAILED — skipping candidato tests');
    alumnoToken = null;
  }

  const editorToken = await login('145B3B9FDS', 'TestEdit2024!');
  console.log('   Editor login: OK');

  // ── 2. candidate-pay: admin gets 403 ──
  console.log('\n2. candidate-pay: admin role blocked');
  const r1 = await request('POST', '/payments/candidate-pay', adminToken, {
    group_exam_id: 1, token: 'fake-token', payment_method_id: 'visa',
  });
  assert(r1.status === 403, `Admin gets 403 (got ${r1.status})`);
  assert(
    r1.data && r1.data.error && r1.data.error.toLowerCase().includes('candidato'),
    'Error mentions candidatos'
  );

  // ── 3. candidate-pay: missing group_exam_id ──
  console.log('\n3. candidate-pay: missing group_exam_id');
  if (alumnoToken) {
    const r2 = await request('POST', '/payments/candidate-pay', alumnoToken, {
      token: 'fake', payment_method_id: 'visa',
    });
    assert(r2.status === 400, `Missing group_exam_id → 400 (got ${r2.status})`);
  } else {
    console.log('  SKIP: No alumno token');
  }

  // ── 4. candidate-pay: missing token ──
  console.log('\n4. candidate-pay: missing token');
  if (alumnoToken) {
    const r3 = await request('POST', '/payments/candidate-pay', alumnoToken, {
      group_exam_id: 99999, payment_method_id: 'visa',
    });
    assert(r3.status === 400, `Missing token → 400 (got ${r3.status})`);
  } else {
    console.log('  SKIP: No alumno token');
  }

  // ── 5. candidate-pay: invalid group_exam_id → 404 ──
  console.log('\n5. candidate-pay: invalid group_exam_id');
  if (alumnoToken) {
    const r4 = await request('POST', '/payments/candidate-pay', alumnoToken, {
      group_exam_id: 999999, token: 'fake-token', payment_method_id: 'visa',
      payer_email: 'test@test.com',
    });
    assert(r4.status === 404, `Invalid group_exam_id → 404 (got ${r4.status})`);
  } else {
    console.log('  SKIP: No alumno token');
  }

  // ── 6. candidate-retake: admin gets 403 ──
  console.log('\n6. candidate-retake: admin role blocked');
  const r5 = await request('POST', '/payments/candidate-retake', adminToken, {
    group_exam_id: 1, token: 'fake-token', payment_method_id: 'visa',
  });
  assert(r5.status === 403, `Admin gets 403 for retake (got ${r5.status})`);

  // ── 7. candidate-retake: invalid group_exam_id ──
  console.log('\n7. candidate-retake: invalid group_exam_id');
  if (alumnoToken) {
    const r6 = await request('POST', '/payments/candidate-retake', alumnoToken, {
      group_exam_id: 999999, token: 'fake-token', payment_method_id: 'visa',
      payer_email: 'test@test.com',
    });
    assert(r6.status === 404, `Invalid group_exam_id → 404 (got ${r6.status})`);
  } else {
    console.log('  SKIP: No alumno token');
  }

  // ── 8. mis-examenes: includes payment fields ──
  console.log('\n8. mis-examenes: payment fields');
  if (alumnoToken) {
    const r7 = await request('GET', '/partners/mis-examenes', alumnoToken);
    assert(r7.status === 200, `mis-examenes returns 200 (got ${r7.status})`);
    if (r7.data && r7.data.exams && r7.data.exams.length > 0) {
      const exam = r7.data.exams.find(e => e.group_id);
      if (exam) {
        assert('requires_payment' in exam, 'Exam has requires_payment field');
        assert('is_paid' in exam, 'Exam has is_paid field');
        assert('certification_cost' in exam, 'Exam has certification_cost field');
        assert(typeof exam.requires_payment === 'boolean', 'requires_payment is boolean');
        assert(typeof exam.is_paid === 'boolean', 'is_paid is boolean');

        // ── 9. check-access: includes payment fields ──
        console.log('\n9. check-access: payment fields');
        const examId = exam.exam_id || exam.id;
        const geid = exam.group_exam_id;
        if (geid) {
          const r8 = await request('GET', `/exams/${examId}/check-access?group_exam_id=${geid}`, alumnoToken);
          assert(r8.status === 200, `check-access returns 200 (got ${r8.status})`);
          assert('requires_payment' in r8.data, 'check-access has requires_payment');
          assert('is_paid' in r8.data, 'check-access has is_paid');
        } else {
          console.log('  SKIP: No group_exam_id on exam');
        }
      } else {
        console.log('  SKIP: No exams with group_id found');
      }
    } else {
      console.log('  SKIP: Alumno has no assigned exams');
    }
  } else {
    console.log('  SKIP: No alumno token');
  }

  // ── 10. my-payments: payment_type field ──
  console.log('\n10. my-payments: payment_type field');
  if (alumnoToken) {
    const r9 = await request('GET', '/payments/my-payments', alumnoToken);
    assert(r9.status === 200, `my-payments returns 200 (got ${r9.status})`);
    if (r9.data && r9.data.payments && r9.data.payments.length > 0) {
      const p = r9.data.payments[0];
      assert('payment_type' in p, 'Payment has payment_type field');
      assert(
        ['voucher', 'certification', 'retake'].includes(p.payment_type),
        `payment_type is valid: ${p.payment_type}`
      );
    } else {
      console.log('  INFO: No payments found for alumno — field check skipped');
      passed++; // Not a failure
    }
  } else {
    console.log('  SKIP: No alumno token');
  }

  // ── 11. Editor cannot use candidate-pay ──
  console.log('\n11. editor: candidate-pay blocked');
  const r10 = await request('POST', '/payments/candidate-pay', editorToken, {
    group_exam_id: 1, token: 'fake', payment_method_id: 'visa',
  });
  assert(r10.status === 403, `Editor gets 403 (got ${r10.status})`);

  // ── 12. No auth → 401/422 ──
  console.log('\n12. candidate-pay: no auth');
  const r11 = await request('POST', '/payments/candidate-pay', null, {
    group_exam_id: 1, token: 'fake', payment_method_id: 'visa',
  });
  assert([401, 422].includes(r11.status), `No auth → 401/422 (got ${r11.status})`);

  // ── Summary ──
  console.log(`\n=== ${ENV.toUpperCase()}: ${passed} passed, ${failed} failed ===\n`);
  if (failed > 0) process.exit(1);
}

main().catch(e => { console.error('FATAL:', e.message); process.exit(1); });

File: MotorUniversalV2/backend/tests/test_candidate_payments.py
"""
Tests del módulo de Pagos por Candidato (Mercado Pago)
======================================================

Cubre:
  A. POST /payments/candidate-pay — Validaciones y control de acceso
  B. POST /payments/candidate-retake — Validaciones y control de acceso
  C. GET  /exams/{id}/check-access — Incluye info de pago
  D. GET  /partners/mis-examenes — Incluye info de pago
  E. Verificar Payment model incluye nuevos campos (group_exam_id, payment_type)

Ejecutar:
    cd backend && python -m pytest tests/test_candidate_payments.py -v --tb=short
    cd backend && python -m pytest tests/test_candidate_payments.py -v -k "candidate_pay"
"""

import pytest
import requests
import time

# ─── Configuración ──────────────────────────────────────────────────────────
DEV_API = "https://evaluaasi-motorv2-api-dev.purpleocean-384694c4.southcentralus.azurecontainerapps.io/api"
ADMIN_USER = "admin"
ADMIN_PASS = "admin123"
ALUMNO_USER = "U343Z793S8"
ALUMNO_PASS = "TestPay2024!"


# ─── Fixtures ───────────────────────────────────────────────────────────────
@pytest.fixture(scope="session")
def api():
    return DEV_API


@pytest.fixture(scope="session")
def admin_token(api):
    """JWT admin con retry para cold-starts."""
    for attempt in range(3):
        try:
            r = requests.post(f"{api}/auth/login", json={
                "username": ADMIN_USER, "password": ADMIN_PASS
            }, timeout=30)
            if r.status_code == 200:
                return r.json()["access_token"]
            if r.status_code in [502, 503]:
                time.sleep(10)
                continue
            pytest.fail(f"Login admin falló: {r.status_code}")
        except requests.exceptions.Timeout:
            if attempt < 2:
                time.sleep(10)
                continue
            pytest.fail("Login timeout")
    pytest.fail("Login admin falló después de 3 intentos")


@pytest.fixture(scope="session")
def alumno_token(api):
    """JWT del candidato (alumno)."""
    for attempt in range(3):
        try:
            r = requests.post(f"{api}/auth/login", json={
                "username": ALUMNO_USER, "password": ALUMNO_PASS
            }, timeout=30)
            if r.status_code == 200:
                return r.json()["access_token"]
            if r.status_code in [502, 503]:
                time.sleep(10)
                continue
            pytest.fail(f"Login alumno falló: {r.status_code}")
        except requests.exceptions.Timeout:
            if attempt < 2:
                time.sleep(10)
                continue
            pytest.fail("Login timeout")
    pytest.fail("Login alumno falló después de 3 intentos")


@pytest.fixture(scope="session")
def admin_headers(admin_token):
    return {
        "Authorization": f"Bearer {admin_token}",
        "Content-Type": "application/json"
    }


@pytest.fixture(scope="session")
def alumno_headers(alumno_token):
    return {
        "Authorization": f"Bearer {alumno_token}",
        "Content-Type": "application/json"
    }


@pytest.fixture(scope="session")
def alumno_exams(api, alumno_headers):
    """Obtener la lista de exámenes asignados al alumno."""
    r = requests.get(f"{api}/partners/mis-examenes", headers=alumno_headers, timeout=20)
    if r.status_code == 200:
        return r.json().get("exams", [])
    return []


@pytest.fixture(scope="session")
def exam_with_payment(alumno_exams):
    """Buscar un examen que requiera pago (si existe)."""
    for exam in alumno_exams:
        if exam.get("requires_payment"):
            return exam
    return None


@pytest.fixture(scope="session")
def exam_without_payment(alumno_exams):
    """Buscar un examen que NO requiera pago."""
    for exam in alumno_exams:
        if not exam.get("requires_payment"):
            return exam
    return None


# ─── A. POST /payments/candidate-pay — Validaciones ─────────────────────────

class TestCandidatePayValidation:
    """Validaciones del endpoint candidate-pay."""

    def test_requires_auth(self, api):
        """Sin JWT retorna 401/422."""
        r = requests.post(f"{api}/payments/candidate-pay",
                          json={"group_exam_id": 1},
                          timeout=15)
        assert r.status_code in [401, 422]

    def test_admin_cannot_use(self, api, admin_headers):
        """Admin no puede usar candidate-pay (solo candidatos)."""
        r = requests.post(f"{api}/payments/candidate-pay",
                          json={
                              "group_exam_id": 1,
                              "token": "fake-token",
                              "payment_method_id": "visa",
                          },
                          headers=admin_headers, timeout=15)
        assert r.status_code == 403
        assert "solo los candidatos" in r.json().get("error", "").lower()

    def test_missing_group_exam_id(self, api, alumno_headers):
        """Sin group_exam_id retorna 400."""
        r = requests.post(f"{api}/payments/candidate-pay",
                          json={
                              "token": "fake-token",
                              "payment_method_id": "visa",
                          },
                          headers=alumno_headers, timeout=15)
        assert r.status_code == 400
        assert "group_exam_id" in r.json().get("error", "").lower()

    def test_missing_token(self, api, alumno_headers):
        """Sin token de tarjeta retorna 400."""
        r = requests.post(f"{api}/payments/candidate-pay",
                          json={
                              "group_exam_id": 99999,
                              "payment_method_id": "visa",
                          },
                          headers=alumno_headers, timeout=15)
        assert r.status_code == 400
        assert "token" in r.json().get("error", "").lower()

    def test_missing_payment_method(self, api, alumno_headers):
        """Sin payment_method_id retorna 400."""
        r = requests.post(f"{api}/payments/candidate-pay",
                          json={
                              "group_exam_id": 99999,
                              "token": "fake-token",
                          },
                          headers=alumno_headers, timeout=15)
        assert r.status_code == 400
        assert "método" in r.json().get("error", "").lower() or "payment_method" in r.json().get("error", "").lower()

    def test_invalid_group_exam_id(self, api, alumno_headers):
        """group_exam_id inexistente retorna 404."""
        r = requests.post(f"{api}/payments/candidate-pay",
                          json={
                              "group_exam_id": 999999,
                              "token": "fake-token",
                              "payment_method_id": "visa",
                              "payer_email": "test@test.com",
                          },
                          headers=alumno_headers, timeout=15)
        assert r.status_code == 404
        assert "no encontrada" in r.json().get("error", "").lower()

    def test_empty_body(self, api, alumno_headers):
        """Body vacío retorna 400."""
        r = requests.post(f"{api}/payments/candidate-pay",
                          json={},
                          headers=alumno_headers, timeout=15)
        assert r.status_code == 400


# ─── B. POST /payments/candidate-retake — Validaciones ──────────────────────

class TestCandidateRetakeValidation:
    """Validaciones del endpoint candidate-retake."""

    def test_requires_auth(self, api):
        """Sin JWT retorna 401/422."""
        r = requests.post(f"{api}/payments/candidate-retake",
                          json={"group_exam_id": 1},
                          timeout=15)
        assert r.status_code in [401, 422]

    def test_admin_cannot_use(self, api, admin_headers):
        """Admin no puede usar candidate-retake (solo candidatos)."""
        r = requests.post(f"{api}/payments/candidate-retake",
                          json={
                              "group_exam_id": 1,
                              "token": "fake-token",
                              "payment_method_id": "visa",
                          },
                          headers=admin_headers, timeout=15)
        assert r.status_code == 403
        assert "solo los candidatos" in r.json().get("error", "").lower()

    def test_missing_group_exam_id(self, api, alumno_headers):
        """Sin group_exam_id retorna 400."""
        r = requests.post(f"{api}/payments/candidate-retake",
                          json={
                              "token": "fake-token",
                              "payment_method_id": "visa",
                          },
                          headers=alumno_headers, timeout=15)
        assert r.status_code == 400
        assert "group_exam_id" in r.json().get("error", "").lower()

    def test_missing_token(self, api, alumno_headers):
        """Sin token retorna 400."""
        r = requests.post(f"{api}/payments/candidate-retake",
                          json={
                              "group_exam_id": 99999,
                              "payment_method_id": "visa",
                          },
                          headers=alumno_headers, timeout=15)
        assert r.status_code == 400
        assert "token" in r.json().get("error", "").lower()

    def test_invalid_group_exam_id(self, api, alumno_headers):
        """group_exam_id inexistente retorna 404."""
        r = requests.post(f"{api}/payments/candidate-retake",
                          json={
                              "group_exam_id": 999999,
                              "token": "fake-token",
                              "payment_method_id": "visa",
                              "payer_email": "test@test.com",
                          },
                          headers=alumno_headers, timeout=15)
        assert r.status_code == 404
        assert "no encontrada" in r.json().get("error", "").lower()

    def test_empty_body(self, api, alumno_headers):
        """Body vacío retorna 400."""
        r = requests.post(f"{api}/payments/candidate-retake",
                          json={},
                          headers=alumno_headers, timeout=15)
        assert r.status_code == 400


# ─── C. GET /partners/mis-examenes — Info de pago ───────────────────────────

class TestMisExamenesPaymentInfo:
    """Verificar que mis-examenes incluye campos de pago."""

    def test_mis_examenes_returns_exams(self, api, alumno_headers, alumno_exams):
        """mis-examenes retorna lista de exámenes."""
        r = requests.get(f"{api}/partners/mis-examenes", headers=alumno_headers, timeout=20)
        assert r.status_code == 200
        data = r.json()
        assert "exams" in data

    def test_exams_include_payment_fields(self, api, alumno_headers, alumno_exams):
        """Cada examen con grupo incluye campos requires_payment, is_paid, certification_cost."""
        if not alumno_exams:
            pytest.skip("Alumno no tiene exámenes asignados")

        for exam in alumno_exams:
            # Solo exámenes con contexto de grupo tendrán estos campos
            if exam.get("group_id"):
                assert "requires_payment" in exam, f"Falta requires_payment en examen {exam.get('id')}"
                assert "is_paid" in exam, f"Falta is_paid en examen {exam.get('id')}"
                assert "certification_cost" in exam, f"Falta certification_cost en examen {exam.get('id')}"
                assert isinstance(exam["requires_payment"], bool)
                assert isinstance(exam["is_paid"], bool)

    def test_exam_with_payment_has_cost(self, exam_with_payment):
        """Si un examen requiere pago, tiene certification_cost > 0."""
        if not exam_with_payment:
            pytest.skip("No hay exámenes con pago requerido en DEV")
        assert exam_with_payment["certification_cost"] > 0

    def test_exam_without_payment_not_required(self, exam_without_payment):
        """Examen sin pagos habilitados tiene requires_payment=False."""
        if not exam_without_payment:
            pytest.skip("No hay exámenes sin pago en DEV")
        assert exam_without_payment["requires_payment"] is False


# ─── D. check-access — Info de pago ─────────────────────────────────────────

class TestCheckAccessPayment:
    """Verificar que check-access incluye campos de pago."""

    def test_check_access_includes_payment_fields(self, api, alumno_headers, alumno_exams):
        """check-access retorna requires_payment, is_paid, certification_cost."""
        if not alumno_exams:
            pytest.skip("Alumno no tiene exámenes asignados")

        exam = alumno_exams[0]
        exam_id = exam.get("exam_id") or exam.get("id")
        group_exam_id = exam.get("group_exam_id")

        if not group_exam_id:
            pytest.skip("Primer examen no tiene group_exam_id")

        r = requests.get(
            f"{api}/exams/{exam_id}/check-access?group_exam_id={group_exam_id}",
            headers=alumno_headers, timeout=20
        )
        assert r.status_code == 200
        data = r.json()

        # Debe incluir los nuevos campos de pago
        assert "requires_payment" in data, f"Falta requires_payment: {data.keys()}"
        assert "is_paid" in data, f"Falta is_paid: {data.keys()}"
        assert isinstance(data["requires_payment"], bool)
        assert isinstance(data["is_paid"], bool)

    def test_check_access_with_payment_required_blocks_exam(self, api, alumno_headers, exam_with_payment):
        """Si requiere pago y no ha pagado, can_take=False."""
        if not exam_with_payment:
            pytest.skip("No hay exámenes con pago requerido en DEV")

        exam_id = exam_with_payment.get("exam_id") or exam_with_payment.get("id")
        group_exam_id = exam_with_payment.get("group_exam_id")

        if not group_exam_id:
            pytest.skip("Examen con pago no tiene group_exam_id")

        r = requests.get(
            f"{api}/exams/{exam_id}/check-access?group_exam_id={group_exam_id}",
            headers=alumno_headers, timeout=20
        )
        assert r.status_code == 200
        data = r.json()
        assert data["requires_payment"] is True

        # Si no ha pagado, can_take debería ser False
        if not data.get("is_paid"):
            assert data.get("can_take") is False, "can_take debería ser False si no ha pagado"


# ─── E. Payment model — Nuevos campos ───────────────────────────────────────

class TestPaymentModelFields:
    """Verificar que el modelo Payment incluye los nuevos campos."""

    def test_my_payments_includes_new_fields(self, api, alumno_headers):
        """GET /my-payments retorna payment_type y group_exam_id en los pagos."""
        r = requests.get(f"{api}/payments/my-payments", headers=alumno_headers, timeout=15)
        assert r.status_code == 200
        data = r.json()
        assert "payments" in data
        # Si hay pagos, verificar campos
        if data["payments"]:
            p = data["payments"][0]
            assert "payment_type" in p, f"Falta payment_type: {list(p.keys())}"
            assert p["payment_type"] in ["voucher", "certification", "retake"], \
                f"payment_type inválido: {p['payment_type']}"


# ─── F. Cross-role access control ────────────────────────────────────────────

class TestCrossRoleAccess:
    """Verificar que los endpoints de candidato no son accesibles por otros roles."""

    def test_responsable_cannot_candidate_pay(self, api):
        """Un responsable no puede usar candidate-pay."""
        # Intentar login como responsable
        r = requests.post(f"{api}/auth/login", json={
            "username": "67SBAFMTQQ",
            "password": "test123"
        }, timeout=30)
        if r.status_code != 200:
            pytest.skip("Responsable 67SBAFMTQQ no disponible en DEV")

        resp_token = r.json()["access_token"]
        r2 = requests.post(
            f"{api}/payments/candidate-pay",
            json={
                "group_exam_id": 1,
                "token": "fake",
                "payment_method_id": "visa",
            },
            headers={
                "Authorization": f"Bearer {resp_token}",
                "Content-Type": "application/json"
            },
            timeout=15
        )
        assert r2.status_code == 403

    def test_editor_cannot_candidate_pay(self, api):
        """Un editor no puede usar candidate-pay."""
        r = requests.post(f"{api}/auth/login", json={
            "username": "editor",
            "password": "editor123"
        }, timeout=30)
        if r.status_code != 200:
            pytest.skip("Editor no disponible en DEV")

        editor_token = r.json()["access_token"]
        r2 = requests.post(
            f"{api}/payments/candidate-pay",
            json={
                "group_exam_id": 1,
                "token": "fake",
                "payment_method_id": "visa",
            },
            headers={
                "Authorization": f"Bearer {editor_token}",
                "Content-Type": "application/json"
            },
            timeout=15
        )
        assert r2.status_code == 403
